common/src/main/java/io/netty/util/concurrent/UnorderedThreadPoolEventExecutor.java

```diff
@@ -84,94 +88,97 @@ public boolean isShuttingDown() {
+    public Future<Void> shutdownGracefully(long quietPeriod, long timeout, TimeUnit unit) {
         // TODO: At the moment this just calls shutdown but we may be able to do something more smart here which
         //       respects the quietPeriod and timeout.
-        shutdown();
+        executor.shutdown();
         return terminationFuture();
     }
 
     @Override
-    public Future<?> terminationFuture() {
+    public Future<Void> terminationFuture() {
         return terminationFuture.asFuture();
     }
 
     @Override
-    protected <V> RunnableScheduledFuture<V> decorateTask(Runnable runnable, RunnableScheduledFuture<V> task) {
-        return runnable instanceof NonNotifyRunnable ?
-                task : new RunnableScheduledFutureTask<>(this, runnable, task);
-    }
-
-    @Override
-    protected <V> RunnableScheduledFuture<V> decorateTask(Callable<V> callable, RunnableScheduledFuture<V> task) {
-        return new RunnableScheduledFutureTask<>(this, callable, task);
-    }
-
-    @Override
-    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
-        return (ScheduledFuture<?>) super.schedule(command, delay, unit);
+    public Future<Void> schedule(Runnable task, long delay, TimeUnit unit) {
+        return (Future<Void>) executor.schedule(task, delay, unit);
     }
 
     @Override
-    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
-        return (ScheduledFuture<V>) super.schedule(callable, delay, unit);
+    public <V> Future<V> schedule(Callable<V> task, long delay, TimeUnit unit) {
+        return (Future<V>) executor.schedule(task, delay, unit);
     }
 
     @Override
-    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit) {
-        return (ScheduledFuture<?>) super.scheduleAtFixedRate(command, initialDelay, period, unit);
+    public Future<Void> scheduleAtFixedRate(Runnable task, long initialDelay, long period, TimeUnit unit) {
+        return (Future<Void>) executor.scheduleAtFixedRate(task, initialDelay, period, unit);
     }
 
     @Override
-    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay, TimeUnit unit) {
-        return (ScheduledFuture<?>) super.scheduleWithFixedDelay(command, initialDelay, delay, unit);
+    public Future<Void> scheduleWithFixedDelay(Runnable task, long initialDelay, long delay, TimeUnit unit) {
+        return (Future<Void>) executor.scheduleWithFixedDelay(task, initialDelay, delay, unit);
     }
 
     @Override
-    public Future<?> submit(Runnable task) {
-        return (Future<?>) super.submit(task);
+    public Future<Void> submit(Runnable task) {
+        return (Future<Void>) executor.submit(task);
     }
 
     @Override
     public <T> Future<T> submit(Runnable task, T result) {
-        return (Future<T>) super.submit(task, result);
+        return (Future<T>) executor.submit(task, result);
```

Comment:
As `Future.executor()` exists I think we need to take special action to ensure the returned `Future` actually return `UnorderedThreadPoolEventExecutor` and not the "inner" executor. Same is true for all the other methods.

Reply:
This is already the case because of how the inner executor decorates tasks. I have added a test for this.